log4j2-extensions/src/main/java/io/confluent/common/logging/log4j2/StructuredLayout.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.confluent.common.logging.log4j2;
+
+import io.confluent.common.logging.LogRecordBuilder;
+import io.confluent.common.logging.LogRecordStructBuilder;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.connect.storage.Converter;
+import org.apache.logging.log4j.core.LogEvent;
+import org.apache.logging.log4j.core.layout.AbstractLayout;
+
+import io.confluent.common.logging.StructuredLogMessage;
+
+import java.util.function.Supplier;
+
+final class StructuredLayout extends AbstractLayout<byte[]> {
+  private static final byte[] EMPTY_BYTES = new byte[0];
+
+  private final String topic;
+  private final Converter converter;
+  private final Supplier<LogRecordBuilder<Struct>> logRecordStructBuilderFactory;
+
+  public byte[] toByteArray(final LogEvent event) {
+    if (event.getMessage().getParameters().length != 1
+        || !(event.getMessage().getParameters()[0] instanceof StructuredLogMessage)) {
+      throw new IllegalArgumentException(
+          "LogEvent must contain a single parameter of type StructuredLogMessage");
+    }
+    final StructuredLogMessage schemaAndValue
+        = (StructuredLogMessage) event.getMessage().getParameters()[0];
+    final Struct logRecord = logRecordStructBuilderFactory.get()
+        .withLoggerName(event.getLoggerName())
+        .withLevel(event.getLevel().name())
+        .withTimeMs(event.getTimeMillis())
+        .withMessageSchemaAndValue(schemaAndValue.getMessage())
+        .build();
+    return converter.fromConnectData(topic, logRecord.schema(), logRecord);
+  }
+
+  public String getContentType() {
+    return "bytes";
+  }
+
+  public byte[] toSerializable(final LogEvent event) {
+    return toByteArray(event);
+  }
+
+  StructuredLayout(
+      final String topic,
```

Comment:
why do we need a topic name for structured logs?

Reply:
Its needed for the avro layout so it can compute a subject name to write the schema under. We should clean it up from this class though. Fixing that.

Reply:
yeah, it feels like a lot of connect-api specifics are leaking into the abstraction. It feels like the format / schema should just be implementation details, and the structured logging interface should just provide the necessary skeleton for us to plug in various implementations, whether that's json, avro, protobuf, it shouldn't really matter.

Reply:
That's pretty much how it's implemented.

The stuff under `common/logging/` is the logging interface. It has no details about the underlying format. It just has an interface that accepts SchemaAndValue as a structured log record, and an implementation that passes it along to the underlying logger. The object that's passed along (SerializableSchemaAndValue) does convert the msg to json, but that's just because some of the logging implementations (e.g. log4j1) force everything into String. There's also some builders there that help the various layouts for the various logging implementations generate the final log records with details about level, time, etc.

The layouts under log4j-extensions and log4j2-extensions is where the actual msg format is decided. This stuff is all pluggable via log4j1/log4j2 configuration.